wayland: Set EGL_SWAP_BEHAVIOR_PRESERVED_BIT on surface create

LiquidityC opened the pull request:
## Description
Set EGL_SWAP_BUFFERS_PRESERVE_BIT when using wayland wm. The working theory is that this was implicitly set on X clients but not on wayland clients.

### Note
Video coding is not my comfort zone so this fix is mostly me guessing on how we'd want to fix this and knowing about one line of code that does fix the problem. Please comment and set me on the right path if this is not how we want to solve this issue.

## Existing Issue(s)
Issue: #7647

Kontrabant commented on src/video/wayland/SDL_waylandopengles.c:
```diff
@@ -41,6 +41,8 @@ int Wayland_GLES_LoadLibrary(SDL_VideoDevice *_this, const char *path)
 
     ret = SDL_EGL_LoadLibrary(_this, path, (NativeDisplayType)data->display, _this->gl_config.egl_platform);
 
+    _this->gl_config.swap_behavior_preserved_bit = 1;
```
This bit is only available if the EGL version >= 1.4, so the current version should be checked before enabling it.

Kontrabant reviewed (changes requested):
See my above comment regarding the implementation.

This still sounds like glReadPixels is doing something strange and reading from a buffer other than the current framebuffer. If this fixes the reproduction example in the linked issue, it's even stranger, as the screenshot is being read from the framebuffer before anything is actually presented, so this flag shouldn't make a difference in that case.

Kontrabant commented on src/video/SDL_egl.c:
```diff
@@ -816,6 +816,10 @@ static int SDL_EGL_PrivateChooseConfig(SDL_VideoDevice *_this, SDL_bool set_conf
         attribs[i++] = _this->egl_data->egl_surfacetype;
     }
 
+    if (_this->gl_config.swap_behavior_preserved_bit) {
```
This bit is only available if the EGL version >= 1.4, so it should be guarded by an #ifdef, and the current version should be checked before enabling it.

LiquidityC replied:
Thanks

Kontrabant commented on src/video/SDL_sysvideo.h:
```diff
@@ -417,6 +417,7 @@ struct SDL_VideoDevice
         int retained_backing;
         int egl_platform;
         int driver_loaded;
+        int swap_behavior_preserved_bit;
```
The EGL data struct in SDL_egl.c is probably a better place for this, as it is EGL specific.

LiquidityC replied:
Moved

LiquidityC commented:
I've addressed your comments. As for the confusion about the fix. I haven't investigated much further and I'm not that experienced with video drivers and rendering.
Compiling the example code with and without this patch does prove that it's doing something because I just get a blank transparent image without it.
The one thing that's changed is that this computer (not sure if it's the same) now uses the vulkan renderer instead of opengl. I just realized. Not sure if that changes anything though.

Kontrabant commented:
I tested your patch, and it segfaults on Nvidia drivers. Mesa doesn't crash, but it never finds a matching EGL config. The EGL_SWAP_BEHAVIOR_PRESERVED_BIT is a bit for the EGL_SURFACE_TYPE attribute parameter. Simply putting it in the array does nothing at best, and crashes in this case.

This patch "fixes" the issue by creating a malformed attribute list causing a matching EGL configuration to never be found, which (when it doesn't crash) causes SDL to skip over the GLES and GL drivers completely in favor of Vulkan or the software renderer.

I pushed a change that tries to enable it properly (it seems to be unsupported on many drivers), so try it with screenshot example code from the linked issue.

LiquidityC commented:
That would explain how it's "working" then. 

I'm traveling for 2 weeks now. I'll test it when I get back.

LiquidityC commented:
Just a ping that I haven't forgotten about this. Just vacation times. I'll be back to normal scheduling in mid August

LiquidityC commented:
When attempting to confirm this problem without my submitted change (and without pulling from main) I'm no longer able to reproduce the issue. This is the case on both laptops where I was seeing the issue. I can't really explain it. I guess some related library or service must have been updated.

We can drop this PR and I will close the related issue.
